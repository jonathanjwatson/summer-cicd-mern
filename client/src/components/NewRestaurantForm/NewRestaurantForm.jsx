import React, { useState } from "react";
import FormInput from "../FormInput/FormInput";
import axios from "axios";

const NewRestaurantForm = ({ getRestaurants }) => {
  const [name, setName] = useState("");
  const [address, setAddress] = useState("");
  const [phone, setPhone] = useState("");
  const [cuisine, setCuisine] = useState("");

  // TODO: When I submit the form
  // 1. Make an API call to send the data.
  // 2. if the call is successful, clear the form fields.
  // 3. If the call is successful, use the prop method to retrieve all restaurants.

  const handleSubmit = (e) => {
    e.preventDefault();
    axios
      .post("/api/restaurants", { name, address, phone, cuisine })
      .then((response) => {
        console.log(response.data);
        setName("");
        setAddress("");
        setPhone("");
        setCuisine("");
        getRestaurants();
      })
      .catch((err) => {
        console.log(err);
      });
  };

  return (
    <form onSubmit={handleSubmit} data-testid="restaurant-form">
      <FormInput value={name} handleChange={setName} name="Name" id="name" />
      <FormInput
        value={address}
        handleChange={setAddress}
        name="Address"
        id="restaurant-address"
      />
      <FormInput
        value={phone}
        handleChange={setPhone}
        name="Phone"
        id="restaurant-phone"
      />
      <FormInput
        value={cuisine}
        handleChange={setCuisine}
        name="Cuisine"
        id="restaurant-cuisine"
      />
      <div className="text-center">
        <button type="submit" className="btn btn-primary">
          Submit
        </button>
      </div>
    </form>
  );
};

export default NewRestaurantForm;
