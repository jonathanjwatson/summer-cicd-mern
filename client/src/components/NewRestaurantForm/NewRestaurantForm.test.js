import React from "react";
import { render, cleanup, fireEvent } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";
import axios from "axios";
import NewRestaurantForm from "./NewRestaurantForm";

afterEach(cleanup);

describe("NewRestaurantForm", () => {
  it("updates the name input value on change", () => {
    const { getByLabelText } = render(<NewRestaurantForm />);
    const nameInput = getByLabelText("Name");
    fireEvent.change(nameInput, { target: { value: "Waffle House" } });
    expect(nameInput.value).toBe("Waffle House");
  });

  it("updates the address input value on change", () => {
    const { getByLabelText } = render(<NewRestaurantForm />);
    const addressInput = getByLabelText("Address");
    fireEvent.change(addressInput, {
      target: { value: "1313 Mockingbird Lane" },
    });
    expect(addressInput.value).toBe("1313 Mockingbird Lane");
  });

  it("updates the phone input value on change", () => {
    const { getByLabelText } = render(<NewRestaurantForm />);
    const addressInput = getByLabelText("Phone");
    fireEvent.change(addressInput, {
      target: { value: "555-555-5555" },
    });
    expect(addressInput.value).toBe("555-555-5555");
  });

  it("updates the cuisine input value on change", () => {
    const { getByLabelText } = render(<NewRestaurantForm />);
    const addressInput = getByLabelText("Cuisine");
    fireEvent.change(addressInput, {
      target: { value: "Seafood" },
    });
    expect(addressInput.value).toBe("Seafood");
  });

  it("should call axios when the form submits", () => {
    // ARRANGE
    axios.post = jest
      .fn()
      .mockImplementationOnce(() => Promise.resolve({ data: true }));
    const { getByTestId } = render(<NewRestaurantForm />);
    const form = getByTestId("restaurant-form");
    // ACT
    fireEvent.submit(form);
    // ASSERT
    expect(axios.post).toHaveBeenCalled();
  });

  it("should call the getRestaurants prop when the form submits", async () => {
    // ARRANGE
    const props = {
      getRestaurants: jest.fn(),
    };
    axios.post = jest
      .fn()
      .mockImplementationOnce(() => Promise.resolve({ data: "banana" }));
    const { getByTestId } = render(<NewRestaurantForm {...props} />);
    const form = getByTestId("restaurant-form");
    // ACT
    await fireEvent.submit(form);
    // ASSERT
    await expect(props.getRestaurants).toHaveBeenCalled();
  });

  it("should wipe the input field values when the form submits", async () => {
    // ARRANGE
    axios.post = jest
      .fn()
      .mockImplementationOnce(() => Promise.resolve({ data: "banana" }));
    const { getByLabelText, getByTestId } = render(<NewRestaurantForm />);
    const form = getByTestId("restaurant-form");
    // ACT
    const nameInput = getByLabelText("Name");
    await fireEvent.change(nameInput, { target: { value: "Waffle House" } });
    await fireEvent.submit(form);
    // ASSERT

    await expect(nameInput.value).toBe("");
  });

  it("matches snapshot", () => {
    const { asFragment } = render(<NewRestaurantForm />);
    expect(asFragment()).toMatchSnapshot();
  });
});
