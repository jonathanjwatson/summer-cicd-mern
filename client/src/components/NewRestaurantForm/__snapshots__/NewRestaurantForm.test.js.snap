// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewRestaurantForm matches snapshot 1`] = `
<DocumentFragment>
  <form
    data-testid="restaurant-form"
  >
    <div
      class="mb-3"
    >
      <label
        class="form-label"
        for="name"
      >
        Name
      </label>
      <input
        class="form-control"
        id="name"
        name="name"
        type="text"
        value=""
      />
    </div>
    <div
      class="mb-3"
    >
      <label
        class="form-label"
        for="restaurant-address"
      >
        Address
      </label>
      <input
        class="form-control"
        id="restaurant-address"
        name="restaurant-address"
        type="text"
        value=""
      />
    </div>
    <div
      class="mb-3"
    >
      <label
        class="form-label"
        for="restaurant-phone"
      >
        Phone
      </label>
      <input
        class="form-control"
        id="restaurant-phone"
        name="restaurant-phone"
        type="text"
        value=""
      />
    </div>
    <div
      class="mb-3"
    >
      <label
        class="form-label"
        for="restaurant-cuisine"
      >
        Cuisine
      </label>
      <input
        class="form-control"
        id="restaurant-cuisine"
        name="restaurant-cuisine"
        type="text"
        value=""
      />
    </div>
    <div
      class="text-center"
    >
      <button
        class="btn btn-primary"
        type="submit"
      >
        Submit
      </button>
    </div>
  </form>
</DocumentFragment>
`;
